// baseplayer/src/types.ts
export interface Rect {
  x: number
  y: number
  width: number
  height: number
}

export interface SequenceBaseItem {
  id?: string | number
  filter?: {
    glsl: string
  }
  crop?: Rect // 裁剪区域（0 ~ 100 百分比）
}

export interface SequenceVideoItem extends SequenceBaseItem {
  type: 'video'
  src: string
  seek: number
  duration: number
  mute?: boolean
}

export interface SequenceImageItem extends SequenceBaseItem {
  type: 'image'
  src: string
  duration: number
}

export interface TransitionItem extends SequenceBaseItem {
  type: 'transition'
  /** 效果 */
  transition?: string
  glsl?: string
  /** 动画曲线 */
  easing?: string
  /** 持续时间 */
  duration: number
}

export interface AudioItem {
  id?: string | number
  type: 'audio'
  src: string
  startTime: number
  duration: number
  fade?: boolean
}

/**
 * 局部滤镜（高斯模糊和马赛克）
 */
export interface PartFilter {
  filterName: string // 'kawase'
  startTime: number
  duration: number
  x: number
  y: number
  width: number
  height: number
}

type SequenceItem = SequenceVideoItem | SequenceImageItem | TransitionItem

export type VideoList = SequenceItem[]

export interface TimelineItem {
  ref: SequenceItem
  startTime: number
  endTime: number
}
