// baseplayer/webpack.config.ts
import path from 'path'
import webpack, { Configuration } from 'webpack'
import HtmlWebpackPlugin from 'html-webpack-plugin'

const rel = (filename: string) => path.resolve(__dirname, filename)

export default (env: any, argv: any) => {
  const entries = {
    index: ['./src/index.ts'],
    demo: ['./example.tsx'],
  }

  const config: Configuration = {
    entry: entries,

    output: {
      path: rel('dist'),
      filename: '[name].js',
      publicPath: '/',
    },

    module: {
      rules: [
        {
          test: /\.tsx?$/,
          use: [
            {
              loader: 'ts-loader',
              options: {
                compilerOptions: {
                  module: 'es2015',
                },
              },
            },
          ],
        },
        {
          test: /\.(mp4|png|jpg|jpeg|png|svg|cur)$/,
          use: [
            'file-loader?name=res/[name]-[hash:6].[ext]',
          ],
          exclude: /icons/,
        },
      ],
    },

    plugins: [
      new HtmlWebpackPlugin({
        filename: 'index.html',
        template: 'template.html',
        inject: true,
        chunks: ['demo'],
      }),
    ],

    resolve: {
      extensions: ['.js', '.ts', '.json', '.tsx'],
      alias: {
        'pixi.js': path.resolve(__dirname, 'pixi.js'),
      },
    },

    devtool: argv.mode === 'development' ? 'inline-source-map' : false,

    devServer: {
      disableHostCheck: true,
    },
  }

  return config
}
